redis_commands.c

```diff
@@ -1818,141 +1818,85 @@ int redis_key_str_arr_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
  * timeout value.  This can handle various SUNION/SUNIONSTORE/BRPOP type
  * commands. */
 static int gen_varkey_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
-                          char *kw, int kw_len, int min_argc, int has_timeout,
+                          char *kw, int kw_len, zend_bool has_timeout,
                           char **cmd, int *cmd_len, short *slot)
 {
-    zval *z_args, *z_ele, ztimeout = {0};
-    HashTable *ht_arr;
-    char *key;
-    int key_free, i, tail;
-    size_t key_len;
-    int single_array = 0, argc = ZEND_NUM_ARGS();
+    zval *argv = NULL, ztimeout = {0}, *zv;
     smart_string cmdstr = {0};
+    uint32_t min_argc;
     short kslot = -1;
-    zend_string *zstr;
+    int single_array;
+    int argc = 0;
 
-    if (argc < min_argc) {
-        zend_wrong_param_count();
-        return FAILURE;
-    }
+    min_argc = has_timeout ? 2 : 1;
 
-    // Allocate args
-    z_args = emalloc(argc * sizeof(zval));
-    if (zend_get_parameters_array(ht, argc, z_args) == FAILURE) {
-        efree(z_args);
-        return FAILURE;
-    }
+    ZEND_PARSE_PARAMETERS_START(min_argc, -1)
+        Z_PARAM_VARIADIC('*', argv, argc)
+    ZEND_PARSE_PARAMETERS_END_EX(return FAILURE);
 
-    // Handle our "single array" case
-    if (has_timeout == 0) {
-        single_array = argc==1 && Z_TYPE(z_args[0]) == IS_ARRAY;
-    } else {
-        single_array = argc==2 && Z_TYPE(z_args[0]) == IS_ARRAY &&
-            (Z_TYPE(z_args[1]) == IS_LONG || Z_TYPE(z_args[1]) == IS_DOUBLE);
+    single_array = argc == min_argc && Z_TYPE(argv[0]) == IS_ARRAY;
+
+    if (has_timeout) {
         if (single_array)
-            ZVAL_COPY_VALUE(&ztimeout, &z_args[1]);
+            ZVAL_COPY_VALUE(&ztimeout, &argv[1]);
+        else
+            ZVAL_COPY_VALUE(&ztimeout, &argv[argc - 1]);
+
+        if (Z_TYPE(ztimeout) != IS_LONG && Z_TYPE(ztimeout) != IS_DOUBLE) {
+            php_error_docref(NULL, E_WARNING, "Timeout must be a long or double");
```

Comment:
Don't we need `zval_dtor(&ztimeout)` here?

Reply:
No, `ZVAL_COPY_VALUE` does not increment the ref count.  It's a shallow copy.

I double-checked to be sure.  If I change `ZVAL_COPY_VALUE` -> `ZVAL_COPY`, PHP will report a leak.